### PR TITLE
Sassafras Protocol v0.3.2

### PR DESCRIPTION
**Pull request by davxy**
Partially Addresses https://github.com/paritytech/polkadot-sdk/issues/41

Delivers:
- ticket structures close to the final form
- experimental integration with [`bandersnatch-vrfs`](https://github.com/w3f/ring-vrf/tree/master/bandersnatch_vrfs)
- ticket ownership claiming via ed25519 challenge

---

Flows into main Sassafras PR https://github.com/paritytech/substrate/pull/11879

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2832151

File: primitives/application-crypto/src/traits.rs [hunk 1]
```diff
@@ -31,7 +31,7 @@ use sp_std::{fmt::Debug, vec::Vec};
 /// Typically, the implementers of this trait are its associated types themselves.
 /// This provides a convenient way to access generic information about the scheme
 /// given any of the associated types.
-pub trait AppCrypto: 'static + Send + Sync + Sized + CryptoType + Clone {
```
Comment (gavofyork) [inline primitives/application-crypto/src/traits.rs:34]: Why?

Comment (davxy) [reply primitives/application-crypto/src/traits.rs:34]: `AppCrypto` is required to be implemented by any `Pair` and thus for the new `bandersnatch::Pair` as well.

Unfortunately `bandersnatch_vrf::SecretKey` embeds the `SecretScalar` ([here](https://github.com/w3f/ring-vrf/blob/4957177a717c7555c8df2869012201017b62e66b/dleq_vrf/src/keys.rs#L86)) AND `SecretScalar` is defined to contain an `UnsafeCell` ([here](https://github.com/w3f/ring-vrf/blob/4957177a717c7555c8df2869012201017b62e66b/ark-secret-scalar/src/lib.rs#L78)). Thus can't be `Sync`.

Nevertheless, `Sync` is not a hard requirement for our internals. So far `Send` seems sufficient for all our use cases

**Comment by davxy** [comment]
(Not ready. Ignore for the bogus review request...)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2839219

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2840160

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2843584

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2843582

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2843583

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2846119

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2846118

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2846387

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2857820

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2857928

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2857927

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2857929

**Pull request merged**